### PR TITLE
Tileset export feature

### PR DESCRIPTION
**Pull request by B-Reif**
When working with tile data, it can be useful to export an image of the Tileset in order to render each tile on the texture exactly once.

The Aseprite beta UI doesn't currently support exporting a Tileset image directly. For now, @dacap shared [a lua script](https://gist.github.com/dacap/9df368ba276a45048dc85f96b23ea818) that will do the same thing. The function in this PR is tested against an image generated by this lua script.

### CODE REVIEW

**Comment by B-Reif** [comment]
I'm not sure how we want to go about resolving the indexed pixels in this case. This may change depending on how we want to factor the pixel format data, per #4. We could store a reference to the required index-resolution data.

**Comment by B-Reif** [comment]
Ok, this PR makes a few new changes:

- Indexed pixels are now resolved during `.image()` instead of during parsing. The pixel indices are still validated to be in the palette's range ahead of time.
- `Pixels` now has a method for converting to `image::Rgba`. It takes a closure for resolving Indexed pixels.
- Grayscale now fully supported with tests, removed assertions.

File: src/pixel.rs [hunk 1]
```diff
@@ -43,13 +48,13 @@ impl Grayscale {
         let alpha = reader.byte()?;
         Ok(Self { value, alpha })
     }
-    pub(crate) fn as_rgba(&self) -> Rgba {
+    pub(crate) fn to_rgba(self) -> Rgba {
```
Comment (alpine-alpaca) [inline src/pixel.rs:51]: Why did you rename this function? Is it because it consumes the Self type? Maybe `into_rgba` would be clearer, as that parallels `into_iter`? Though, I'm not aware of a clear convention.

File: src/pixel.rs [hunk 2]
```diff
@@ -28,6 +30,9 @@ impl Rgba {
             alpha,
         })
     }
+    fn as_image_rgba(&self) -> image::Rgba<u8> {
```
Comment (alpine-alpaca) [inline src/pixel.rs:33]: Btw, I prefer to have an extra newline between definitions. I would like to just have it enforced by `rustfmt` but there doesn't seem to be an existing configuration option.

File: src/pixel.rs [hunk 3]
```diff
@@ -153,33 +158,21 @@ impl Pixels {
...
-    pixels: &[Indexed],
-    palette: &ColorPalette,
-    transparent_color_index: u8,
-    layer_is_background: bool,
-) -> Result<Vec<Rgba>> {
-    pixels
-        .iter()
-        .map(|px| resolve_indexed(px, palette, transparent_color_index, layer_is_background))
-        .collect()
+    pub(crate) fn clone_as_image_rgba<F>(&self, index_resolver: F) -> Vec<image::Rgba<u8>>
```
Comment (alpine-alpaca) [inline src/pixel.rs:162]: Hm, I wonder why we need our own `Rgba` type. I think we could just reuse/re-export the one from `image` since that will remain a required dependency. Then in this function you could return a `std::borrow::Cow` so that we don`t need to copy in case the image data is already in rgba.

File: src/file.rs [hunk 4]
```diff
@@ -209,6 +217,16 @@ impl AsepriteFile {
         &self.tilesets
     }
 
+    /// Construct the image of each tile in the [Tileset].
+    /// The image has width equal to the tile width and height equal to (tile_height * tile_count).
+    pub fn tileset_image(&self, tileset_id: &TilesetId) -> Option<RgbaImage> {
```
Comment (alpine-alpaca) [inline src/file.rs:222]: I think this should return a result so that we can return an error message in case of failure.
Afaict, the only error case that isn't covered by validation is an invalid tileset id?

File: src/file.rs [hunk 5]
```diff
@@ -233,29 +251,34 @@ impl AsepriteFile {
...
     }
 
+    fn resolve_indexed_pixel(
+        &self,
+        layer_is_background: bool,
+        pixel: &pixel::Indexed,
+    ) -> pixel::Rgba {
+        let palette = self.palette.as_ref().expect("Expected a palette present when resolving indexed image. Should have been caught in validation");
+        let transparent_color_index = self.pixel_format.transparent_color_index().expect("Indexed tilemap pixels in non-indexed pixel format. Should have been caught in validation");
+        pixel
```
Comment (alpine-alpaca) [inline src/file.rs:261]: Since this function is called per-pixel it probably shouldn't have so many failure points. These first two should always be valid after the first invocation of the function. So, maybe we can make this a standalone function (i.e., no `self` argument) and pass in the palette reference and transparent color index as arguments. Then do the checks where you construct index resolver closure below.

File: src/tileset.rs [hunk 6]
```diff
@@ -134,6 +137,37 @@ impl Tileset {
         self.external_file.as_ref()
     }
 
+    pub(crate) fn image(&self, image_pixels: Vec<image::Rgba<u8>>) -> RgbaImage {
```
Comment (alpine-alpaca) [inline src/tileset.rs:140]: Small nit. Since this is an internal function with the same name as a global API function, it would be nice if this would have a more internal-sounding name. E.g., more along the lines of `tileset_image_from_pixels` (not great). Or even `tilemap_image`. Maybe you can think of something better.

File: src/pixel.rs (hunk 1 above)
Comment (B-Reif) [reply src/pixel.rs:51]: It looks like "into" is the correct one in this case: https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv

File: src/file.rs (hunk 4 above)
Comment (B-Reif) [reply src/file.rs:222]: Right, since this function is exposed to the end-user, they could pass in any unknown tileset id.

File: src/file.rs (hunk 5 above)
Comment (B-Reif) [reply src/file.rs]: The closure would still need to be constructed conditionally to avoid unwrapping stuff when the closure's not needed. I briefly tried a "closure factory" signature (e.g `|| -> |px| -> rgba`) and that caused some lifetime issues. I decided to just have `Pixels` accept the optional arguments, and in the Indexed case, do the unwrapping, and then iterate.

File: src/tileset.rs (hunk 6 above)
Comment (B-Reif) [reply src/tileset.rs]: I decided on `write_to_image` to match similar internal functions that return `RgbaImage`.

**Comment by alpine-alpaca** [comment]
Do you think this is ready for review/merge?

**Comment by B-Reif** [comment]
> Do you think this is ready for review/merge?

Yes, I think so.

**Comment by B-Reif** [comment]
@alpine-alpaca Are there any other changes you'd like here?

**Comment by alpine-alpaca** [comment]
This is included in the merged PR, right?

**Comment by alpine-alpaca** [comment]
Handled by #9
